src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IAsyncEnumerableOfTConverter.cs

```diff
@@ -0,0 +1,144 @@
+        {
+            IAsyncEnumerator<TElement> enumerator;
+            ValueTask<bool> moveNextTask;
+
+            if (state.Current.AsyncEnumerator is null)
+            {
+                enumerator = value.GetAsyncEnumerator(state.CancellationToken);
+                moveNextTask = enumerator.MoveNextAsync();
+                // we always need to attach the enumerator to the stack
+                // since it will need to be disposed asynchronously.
+                state.Current.AsyncEnumerator = enumerator;
+            }
+            else
+            {
+                Debug.Assert(state.Current.AsyncEnumerator is IAsyncEnumerator<TElement>);
+                enumerator = (IAsyncEnumerator<TElement>)state.Current.AsyncEnumerator;
+
+                if (state.Current.AsyncEnumeratorIsPendingCompletion)
+                {
+                    // converter was previously suspended due to a pending MoveNextAsync() task
+                    Debug.Assert(state.PendingTask is Task<bool> && state.PendingTask.IsCompleted);
+                    moveNextTask = new ValueTask<bool>((Task<bool>)state.PendingTask);
+                    state.Current.AsyncEnumeratorIsPendingCompletion = false;
+                    state.PendingTask = null;
+                }
+                else
+                {
+                    // converter was suspended for a different reason;
+                    // the last MoveNextAsync() call can only have completed with 'true'.
+                    moveNextTask = new ValueTask<bool>(true);
+                }
+            }
+
+            JsonConverter<TElement> converter = GetElementConverter(ref state);
+
+            // iterate through the enumerator while elements are being returned synchronously
+            for (; moveNextTask.IsCompleted; moveNextTask = enumerator.MoveNextAsync())
+            {
+                if (!moveNextTask.Result)
+                {
+                    return true;
+                }
+
+                if (ShouldFlush(writer, ref state))
+                {
+                    return false;
+                }
+
+                TElement element = enumerator.Current;
+                if (!converter.TryWrite(writer, element, options, ref state))
+                {
+                    return false;
+                }
+            }
+
+            // we have a pending MoveNextAsync() call;
+            // wrap inside a regular task so that it can be awaited multiple times;
+            // mark the current stackframe as pending completion.
+            Debug.Assert(state.PendingTask is null);
+            state.PendingTask = moveNextTask.AsTask();
+            state.Current.AsyncEnumeratorIsPendingCompletion = true;
+            return false;
+        }
+
+        internal override bool OnTryRead(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options, ref ReadStack state, out TAsyncEnumerable value)
```

Comment:
nit: we usually have the read method first.

Reply:
Good point. These used to throw `NotSupportedException` which is why I had them moved to the bottom.